test/Dialect/Comb/errors.mlir

```diff
@@ -11,3 +11,17 @@ hw.module @err(in %a: i1, in %b: i1) {
   // expected-error @+1 {{Truth tables support a maximum of 63 inputs}}
   %0 = comb.truth_table %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b, %a, %b -> [false]
 }
+
+// -----
+
+hw.module @err(in %a: i0) {
+  // expected-error @+1 {{op replicate must produce integer multiple of operand}}
+  %0 = comb.replicate %a : (i0) -> i16
+}
+
+// -----
+
+hw.module @err(in %a: i0) {
+  // expected-error @+1 {{from bit too large for input}}
+  %0 = comb.extract %a from 0 : (i0) -> i0
+}
```

Comment:
I'm wondering if this is the behavior you'd expect. Intuitively you can replace `comb.extract %a from 0 : (T) -> T` with just `%a`, since the extract is a no-op. That would suggest you can do the same for `T = i0`, which makes this error somewhat unintuitive. I think `comb.extract` should be valid as you've written it. Maybe we should instead check if `lowBit >= 0` and `lowBit + resultWidth <= inputWidth`, which would allow this case, and also `comb.extract %a from 42 : (i42) -> i0`.

Reply:
Do we want to allow `comb.extract %a from 100000 : (i16)-> i0` then? Because conceptually it is the same kind of extraction outside the bounds of the integer. Intuitively it is taking bits where the original value has none, which is only fine if you are taking zero bits.

Reply:
Let me try to approach this from a more mathematical perspective: Let's consider the operation `comb.extract %a from x : (iy) -> iz` where x, y, z are integers >= 0. Let's decompose the input type to the interval of bit indices `I = [0,y)`, and the output type to bit indices `O = [x,x+z)`. Then, the minimal constraint we need for the op to work is `O \subset I` (otherwise, we'd need to define what should be returned for non `i0` OOB accesses). If `z = 0`, then O is the empty set/interval, which is a subset of any other set; thus, it wouldn't matter what x and y are for the operation to be valid (this is what @Moxinilian points out). In the current implementation, we have the additional constraint that `x \element I` which cannot be satisfied if `y = 0` since I is the empty set. @fabianschuiki is suggesting that this constraint should be changed to `x \element (I \union [y, y])`.

I think we shouldn't drop the second constraint entirely because I can't see a case where `comb.extract %a from 100000 : (i16)-> i0` would ever be useful but is more likely a symptom of a bug somewhere that can be caught by the second constraint. In the end, it's a question of definition, and which one we choose is a question of what is more practical to work with in compiler passes. I suppose the extra flexibility @fabianschuiki proposes can be useful when working with module parameters (e.g., if a parameter N is used to extract the N LSB/MSB from an integer).

Reply:
`comb.extract %a from 100000 : (i16)-> i0` should be illegal.  It is very reasonable that in the IR, indexes are legal.  If a front end language wants to have operations which allow "over-the-limit" extractions (like firrtl does), that's fine, but that complexity shouldn't impact comb.

The alternate argument is that parameters will make this hard to enforce.

Reply:
Or put alternately, ops should not encode an implicit extension.  So if `comb.extract %a from 100000 : (i16)-> i0` is legal, that implies the op always implicitly extends the argument sufficiently to perform the extraction.  This makes analysis much harder.  Lots of bugs in FIRRTL come from places where people forgot that the ops have implicit extension and we are slowly reworking the dialect to rectify that.

Reply:
Extracting an i0 from bit 10000 is not an implicit extension though. The property "all extracted bits are defined" is satisfied even when no extension is carried out. But in any case, the verifier I implemented will not accept it, at least for now.